### ISSUE TITLE
Best practice for Recipe and sources in the same repo

### ISSUE THREAD

**Issue by wpalfi** [conan-io/conan#2141]
In 
http://docs.conan.io/en/latest/packaging/package_repo.html
you recommend to use the exports_sources attribute instead of source(). But in 
http://docs.conan.io/en/latest/reference/conanfile/attributes.html#url
you suggest:

> Of course, if you have the conanfile inside your library source, you can point to it, and afterwards use the url in your source() method.

I just read the second solution by accident after implementing the first one. In my opinion the second one has some advantages:
- easier to understand because very similar to external recipe solution
- clean checkout avoids problems

Wouldn't it be better to promote the second solution in the docs?

**Comment by memsharded** [comment]
Using the ``source()`` method when the conanfile.py is inside the same repo has a chicken and egg problem. You need to checkout the sources from a git repo, a certain branch. But that might not exists at all, until you create it. So the flow is a bit weird, you need to do changes, push them to the git remote, then do a ``conan create`` that will clone & checkout the sources. When you already have locally the source code, just besides the conanfile.py, because they live in the same repo.

Both approaches are possible, but the ``exports_sources`` one seems to be more understandable. I am willing to hear other experiences and feedback regarding this!

**Comment by mpdelbuono** [comment]
I have also been using `exports_sources` for the case where conanfile.py is in the same repo. While I toyed with the idea of setting the URL and using the `source()` method, it seemed unnecessarily complicated and confusing when compared to `exports_sources` for minimal benefit. 

That being said, I _do_ recognize the advantages you listed. Like many things, I view it as a tradeoff, and in my particular case, I decided against the `source()` method because I could not find that the benefits outweighed the costs. However, I think this can be highly dependent upon your situation.
